### ISSUE TITLE
Multi-Node Simulation Support

### ISSUE THREAD

**Issue by Nandika-A** [ContriHUB/Congestion-Control-Simulator#5]
# 🚀Feature Request 

### Description  
The current simulator is designed to simulate network scenarios with only two nodes. Introducing support for multi-node network topologies would allow for more complex and realistic simulations, enabling users to model larger and more diverse network environments.

### Why is this feature important to you?  
In real-world networks, multiple nodes interact with one another, leading to various challenges in congestion control and network performance. 

### Describe the solution you'd like  
Implement functionality that allows users to define and simulate networks with multiple nodes. This includes designing a system where packets can flow between various nodes, and congestion control algorithms can operate on a node-to-node basis. The user should be able to configure the number of nodes and the connections between them, and the simulator should display how congestion propagates through the network.

**Comment by Gprakhar141** [comment]
Thanks for assigning, will start working on it right away.

**Comment by Gprakhar141** [comment]
I have implemented support logic for multiple nodes and modelled the nodes and connection between them using graphs and have stored the connections between the nodes in the form of a matrix.
I have used react states to store number of nodes and connections between them.
A state for nodes is kept in which all the info(cwnd, ssthresh, sent_packets, lost_packets and ack) is stored in the form of array of js objects.
I haven't modified the logic behind any algo just optimised them to handle multiple node scenarios.

Kindly ping me if I'm right on it.

**Comment by Gprakhar141** [comment]
![Screenshot 2024-10-10 153641](https://github.com/user-attachments/assets/442b41a1-582d-4674-9cdb-eae9b493bae1)

**Comment by Gprakhar141** [comment]
Should I continue and raise my PR or anything else is there I am missing???

**Comment by Gprakhar141** [comment]
`//Code Snippets for the logic part
```
const handleClick = () => {
    if (!isConnected(selectedNodeId, selectedDestNodeId)) {
      alert(`No connection exists between Node ${selectedNodeId} and Node ${selectedDestNodeId}`);
      return;
    }

    setNodes(prevNodes => {
      const updatedNodes = [...prevNodes];
      const currentNode = { ...updatedNodes[selectedNodeId] };

      if (currentNode.lost.length > 0) {
        currentNode.ack = currentNode.lost[0]; // Acknowledge the lost packet
      } else {
        currentNode.ack = currentNode.sent[currentNode.sent.length - 1] + 1; // ACK for the last sent packet
      }

      updatedNodes[selectedNodeId] = currentNode;
      return updatedNodes;
    });
  };

  const handleNext = () => {
    if (!isConnected(selectedNodeId, selectedDestNodeId)) {
      alert(`No connection exists between Node ${selectedNodeId} and Node ${selectedDestNodeId}`);
      return;
    }

    setNodes(prevNodes => {
      const updatedNodes = [...prevNodes];
      const currentNode = { ...updatedNodes[selectedNodeId] };

      if (currentNode.cwnd < currentNode.ssthresh && currentNode.lost.length === 0) {
        // Increase cwnd
        currentNode.cwnd += 1;
        currentNode.sent = Array.from({ length: currentNode.cwnd + 1 }, (v, i) => currentNode.ack + 1 + i);
      } else if (currentNode.lost.length !== 0) {
        // Handle lost packets
        let last = currentNode.sent[currentNode.sent.length - 1];
        currentNode.cwnd += 1;
        currentNode.sent = [...currentNode.lost]; // Reset sent to lost packets
        if (currentNode.sent.length <= currentNode.cwnd) {
          currentNode.sent = currentNode.sent.concat(Array.from({ length: currentNode.cwnd - currentNode.sent.length + 1 }, (v, i) => last + 1 + i));
        }
        currentNode.lost = [];
      } else {
        // If we reached the threshold
        currentNode.cwnd = 1;
        currentNode.ssthresh = Math.floor(currentNode.ssthresh / 2);
        currentNode.sent = Array.from({ length: currentNode.cwnd + 1 }, (v, i) => currentNode.ack + 1 + i);
      }

      updatedNodes[selectedNodeId] = currentNode;
      return updatedNodes;
    });
  };
```
`

**Comment by Gprakhar141** [comment]
Code Snippets
`
const [nodes, setNodes] = useState([]);//This state will contain all the info(cwnd, ssthresh, sent , lost, ack) for each node
  //in the form of array of js objects
  const [connections, setConnections] = useState([]); // Connection matrix (bi-directional)
  const [numNodes, setNumNodes] = useState(0); // Number of nodes to be created
  const [newConnection, setNewConnection] = useState({ from: 0, to: 0 }); // For creating connections
  const [selectedNodeId, setSelectedNodeId] = useState(0); // Selected source node
  const [selectedDestNodeId, setSelectedDestNodeId] = useState(1); // Selected destination node
  const [lost_pkt, setLost_pkt] = useState(0); // Packet to be marked as lost
`

**Comment by Gprakhar141** [comment]
`
```
// Add a connection between two nodes
  const addConnection = () => {
    const { from, to } = newConnection;
    if (
      from >= 0 && from < nodes.length &&
      to >= 0 && to < nodes.length &&
      from !== to &&
      !connections.some(conn => (conn[0] === from && conn[1] === to) || (conn[1] === from && conn[0] === to))
    ) {
      setConnections([...connections, [from, to]]);
    }
  };
```
`

**Comment by Gprakhar141** [comment]
`
```
//To simulate packet tarnsfer for the current window
  const handleClick = () => {
    if (!isConnected(selectedNodeId, selectedDestNodeId)) {
      alert(`No connection exists between Node ${selectedNodeId} and Node ${selectedDestNodeId}`);
      return;
    }

    setNodes(prevNodes => {
      const updatedNodes = [...prevNodes];
      const currentNode = { ...updatedNodes[selectedNodeId] };

      if (currentNode.lost.length > 0) {
        currentNode.ack = currentNode.lost[0]; // Acknowledge the lost packet
      } else {
        currentNode.ack = currentNode.sent[currentNode.sent.length - 1] + 1; // ACK for the last sent packet
      }

      updatedNodes[selectedNodeId] = currentNode;
      return updatedNodes;
    });
  };
```
`

**Comment by Nandika-A** [comment]
Hi @Gprakhar141! The code review is done only once the PR is raised. So raise the PR as soon as you complete your work and I will provide direct reviews over the changes in the files in the PR. Code snippets aren't required to be sent in the issue comments. Take care from the next time and raise the PR.

**Comment by Nandika-A** [comment]
Closed as completed via #6
